desc/integrals/bounce_integral.py

```diff
@@ -1,26 +1,686 @@
+from numpy.fft import rfft2
 from orthax.legendre import leggauss
 
 from desc.backend import jnp
+from desc.integrals.basis import ChebyshevBasisSet, FourierChebyshevBasis
 from desc.integrals.bounce_utils import (
     _bounce_quadrature,
     _check_bounce_points,
+    _check_interp,
     _set_default_plot_kwargs,
     bounce_points,
+    get_alpha,
     get_pitch_inv,
     interp_to_argmin,
     plot_ppoly,
 )
-from desc.integrals.interp_utils import polyder_vec
+from desc.integrals.interp_utils import interp_rfft2, irfft2_non_uniform, polyder_vec
 from desc.integrals.quad_utils import (
     automorphism_sin,
+    bijection_from_disc,
     get_quadrature,
     grad_automorphism_sin,
 )
 from desc.io import IOAble
-from desc.utils import errorif, setdefault, warnif
+from desc.utils import (
+    atleast_nd,
+    check_posint,
+    errorif,
+    flatten_matrix,
+    setdefault,
+    warnif,
+)
+
+
+def _transform_to_desc(grid, f, is_reshaped=False):
+    """Transform to DESC spectral domain.
+
+    Parameters
+    ----------
+    grid : Grid
+        Tensor-product grid in (θ, ζ) with uniformly spaced nodes in
+        (2π × 2π) poloidal and toroidal coordinates.
+    f : jnp.ndarray
+        Function evaluated on ``grid``.
+
+    Returns
+    -------
+    a : jnp.ndarray
+        Shape (..., grid.num_theta // 2 + 1, grid.num_zeta)
+        Complex coefficients of 2D real FFT of ``f``.
+
+    """
+    if not is_reshaped:
+        f = grid.meshgrid_reshape(f, "rtz")
+    # real fft over poloidal since usually m > n
+    a = rfft2(f, axes=(-1, -2), norm="forward")
+    assert a.shape[-2:] == (grid.num_theta // 2 + 1, grid.num_zeta)
+    return a
+
+
+def _transform_to_clebsch(grid, desc_from_clebsch, B, zeta_0=0.0, is_reshaped=False):
+    """Transform to Clebsch spectral domain.
+
+    Parameters
+    ----------
+    grid : Grid
+        Tensor-product grid in (ρ, θ, ζ) with uniformly spaced nodes in
+        (2π × 2π) poloidal and toroidal coordinates.
+        Note that below shape notation defines
+        L = ``grid.num_rho``, m = ``grid.num_theta``, and n = ``grid.num_zeta``.
+        Preferably power of 2 for ``m`` and ``n``.
+    desc_from_clebsch : jnp.ndarray
+        Shape (L, M, N, 3).
+        DESC coordinates (ρ, θ, ζ) sourced from the Clebsch coordinates
+        ``FourierChebyshevBasis.nodes(M,N,L,domain=(0,2*jnp.pi))``.
+    B : jnp.ndarray
+        |B| evaluated on ``grid``.
+    zeta_0 : float
+        ζ coordinate to begin branch cut of (α, ζ) spline.
+        Default is 0.
+
+    Returns
+    -------
+    T, B : (FourierChebyshevBasis, FourierChebyshevBasis)
+        Spectral coefficients of θ(α, ζ) and |B|(α, ζ).
+
+    """
+    domain = (zeta_0, zeta_0 + 2 * jnp.pi)
+    M, N = desc_from_clebsch.shape[-3:-1]
+    # θ is computed on the optimal nodes in Clebsch space,
+    # which is a tensor product node set in Clebsch space.
+    T = FourierChebyshevBasis(f=desc_from_clebsch[..., 1], domain=domain)
```

Comment:
The only thing holding back this PR right now is that the function f has some tiny non-smooth part around that breaks convergence of `T`. Not sure what the issue is. B  is fine, so this still must be some issue with the coordinate mapping. or perhaps something is weird about the shifted circle equilibrium we are using.

Reply:
https://github.com/PlasmaControl/DESC/pull/1254#issuecomment-2359927383

Reply:
note to self: add zeta_0 to get_alphas